expertsystem/types.py

```diff
@@ -0,0 +1,137 @@
+"""Definitions used internally for typing and signatures."""
```

Comment:
These are all solving related types (except Scalar maybe). I would propose to move this into the solving module, because solving actually defines these types